graspologic/embed/n2v.py

```diff
@@ -0,0 +1,402 @@
+                else:
+                    prev = walk[-2]
+                    next = current_neighbors[
+                        _alias_draw(
+                            alias_edges[(prev, current)][0],
+                            alias_edges[(prev, current)][1],
+                        )
+                    ]
+                    walk.append(next)
+            else:
+                break
+
+        return walk
+
+    @staticmethod
+    def _get_walk_length_interpolated(
+        degree: int, percentiles: np.ndarray, max_walk_length: int
+    ):
+        """
+        Given a node's degree, determine the length of a walk that should be used. If the degree is less than the
+        first element of the percentiles list, default the walk length to 1. Otherwise, if the degree is greater
+        than the last element of the list, default it to the max_walk_length. If it falls in the middle, do a linear
+        interpolation to decide the length of the walk.
+        """
+
+        new_walk_length = None
+
+        for i, percentile in enumerate(percentiles):
+            # if we are below the first percentile in the list, default to a walk length of 1
+            if i == 0 and degree < percentile:
+                return 1
+
+            # otherwise, find which bucket we are going to be in.
+            if degree <= percentile:
+                new_walk_length = max_walk_length * ((i * 0.1) + 0.2)
+                break
+
+        # the degree is above the last percentile
+        if not new_walk_length:
+            new_walk_length = max_walk_length
+
+        # a walk length of 0 is invalid but can happen depending on the percentiles used
+        if new_walk_length < 1:
+            new_walk_length = 1
+
+        return math.floor(new_walk_length)
+
+    def _simulate_walks(
+        self,
+        num_walks: int,
+        walk_length: int,
+        interpolate_walk_lengths_by_node_degree: bool = False,
+    ):
+        """
+        Repeatedly simulate random walks from each node.
+        """
+        graph = self.graph
+        walks = []
+        nodes = list(graph.nodes())
+
+        degree_percentiles: Optional[np.ndarray] = None
+        if interpolate_walk_lengths_by_node_degree:
+            degree_percentiles = np.percentile(
+                [degree for _, degree in graph.degree()], [x for x in range(20, 90, 10)]
+            )
+
+        for walk_iteration in range(num_walks):
+            logging.info(
+                "Walk iteration: " + str(walk_iteration + 1) + "/" + str(num_walks)
+            )
+            random.shuffle(nodes)
+            for node in nodes:
+                walks.append(
+                    self.node2vec_walk(
+                        walk_length=walk_length,
+                        start_node=node,
+                        degree_percentiles=degree_percentiles,
+                    )
+                )
+
+        return walks
+
+    def _get_alias_edge(self, source: Any, destination: Any):
```

Comment:
what is an alias edge?

Reply:
i didnt understand any of the alias terminology until reading https://lips.cs.princeton.edu/the-alias-method-efficient-sampling-with-many-discrete-outcomes/

may be worth having that citation elsewhere so it makes sense to new reader

Reply:
I still don't understand any of the terminology - I am thinking this whole thing needs a revamp in 0.2, if only so you don't have to read a paper to understand what the code is doing first.